File: course_3/module_2/programming_assignment_2/solution_1.py
class UnionFind:
    def __init__(self, size):
        # Initialize the Union-Find structure with the given size.
        # Each node is initially its own parent, and the rank is set to 0.
        if size <= 0:
            raise ValueError("Size of UnionFind must be positive.")
        self.parent = list(range(size))
        self.rank = [0] * size

    def find(self, node):
        # Find the root of the node using path compression.
        # Path compression flattens the structure for faster future queries.
        if self.parent[node] != node:
            self.parent[node] = self.find(self.parent[node])  # Path compression.
        return self.parent[node]

    def union(self, node1, node2):
        # Perform the union operation by rank.
        # Connects two subsets into a single subset.
        root1 = self.find(node1)
        root2 = self.find(node2)

        if root1 != root2:
            if self.rank[root1] > self.rank[root2]:
                self.parent[root2] = root1
            elif self.rank[root1] < self.rank[root2]:
                self.parent[root1] = root2
            else:
                self.parent[root2] = root1
                self.rank[root1] += 1


class Clustering:
    def __init__(self, nodes, edges):
        # Initialize the clustering algorithm with the number of nodes and edges.
        if nodes <= 0:
            raise ValueError("Number of nodes must be positive.")
        if not edges:
            raise ValueError("Edge list cannot be empty.")
        self.nodes = nodes
        self.edges = edges

    def compute_max_spacing(self, k):
        # Compute the maximum spacing of k-clustering.
        if k > self.nodes or k <= 0:
            raise ValueError(f"Invalid number of clusters: {k}. Must be between 1 and {self.nodes}.")

        # Sort the edges by their cost in ascending order.
        self.edges.sort(key=lambda edge: edge[2])  # Edge is now a tuple (node1, node2, cost).
        uf = UnionFind(self.nodes)
        num_clusters = self.nodes

        # Process edges until the number of clusters equals k.
        for node1, node2, cost in self.edges:
            if uf.find(node1) != uf.find(node2):  # Check if nodes are in different clusters.
                if num_clusters == k:
                    # Return the cost of the next edge that connects two different clusters.
                    return cost
                uf.union(node1, node2)  # Merge the clusters.
                num_clusters -= 1

        raise RuntimeError("Unable to determine max spacing. Check input data.")


def read_input(file_path):
    """Reads the input file and parses the nodes and edges."""
    # Open the input file and read its contents line by line.
    with open(file_path, 'r') as file:
        lines = file.readlines()

    # The first line contains the number of nodes.
    num_nodes = int(lines[0].strip())
    edges = []

    # Parse the rest of the lines as edges.
    for line in lines[1:]:
        node1, node2, cost = map(int, line.strip().split())
        edges.append((node1 - 1, node2 - 1, cost))  # Convert to 0-based index.

    return num_nodes, edges


def main():
    # File path to the input data.
    file_path = "clustering1.txt"

    try:
        # Read the input data from the specified file.
        nodes, edges = read_input(file_path)
        clustering = Clustering(nodes, edges)

        # Compute the maximum spacing for 4 clusters.
        max_spacing = clustering.compute_max_spacing(k=4)

        # Print the result to the console.
        print(f"The maximum spacing of a 4-clustering is: {max_spacing}")
    except FileNotFoundError:
        # Handle the case where the input file is not found.
        print(f"Error: File '{file_path}' not found.")
    except Exception as e:
        # Handle any other unexpected errors.
        print(f"Error: {e}")


if __name__ == "__main__":
    # Entry point of the script.
    main()

File: course_3/module_2/programming_assignment_2/solution_1.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
	"strconv"
	"strings"
)

// Edge represents an edge in the graph with two nodes and a cost.
type Edge struct {
	Node1 int // The first node of the edge.
	Node2 int // The second node of the edge.
	Cost  int // The cost or weight of the edge.
}

// UnionFind represents the union-find (disjoint-set) structure.
type UnionFind struct {
	Parent []int // Array to track the parent of each node.
	Rank   []int // Array to track the rank (depth) of each set.
}

// NewUnionFind initializes a new UnionFind structure with the specified size.
func NewUnionFind(size int) *UnionFind {
	parent := make([]int, size)
	rank := make([]int, size)
	for i := 0; i < size; i++ {
		parent[i] = i // Initially, each node is its own parent.
	}
	return &UnionFind{Parent: parent, Rank: rank}
}

// Find finds the root of the node using path compression for efficiency.
func (uf *UnionFind) Find(node int) int {
	if uf.Parent[node] != node {
		uf.Parent[node] = uf.Find(uf.Parent[node]) // Recursively find the root and compress the path.
	}
	return uf.Parent[node]
}

// Union unions two subsets based on their rank to maintain a balanced tree structure.
func (uf *UnionFind) Union(node1, node2 int) {
	root1 := uf.Find(node1) // Find the root of the first node.
	root2 := uf.Find(node2) // Find the root of the second node.

	if root1 != root2 {
		if uf.Rank[root1] > uf.Rank[root2] {
			uf.Parent[root2] = root1 // Attach the smaller tree under the larger tree.
		} else if uf.Rank[root1] < uf.Rank[root2] {
			uf.Parent[root1] = root2
		} else {
			uf.Parent[root2] = root1 // If ranks are equal, attach one tree and increase the rank.
			uf.Rank[root1]++
		}
	}
}

// Clustering represents the clustering algorithm.
type Clustering struct {
	Nodes int    // Number of nodes in the graph.
	Edges []Edge // List of edges in the graph.
}

// NewClustering initializes a new Clustering instance with nodes and edges.
func NewClustering(nodes int, edges []Edge) *Clustering {
	return &Clustering{Nodes: nodes, Edges: edges}
}

// ComputeMaxSpacing computes the maximum spacing of k-clustering.
func (c *Clustering) ComputeMaxSpacing(k int) (int, error) {
	// Validate the value of k.
	if k <= 0 || k > c.Nodes {
		return 0, fmt.Errorf("invalid value of k: must be between 1 and %d", c.Nodes)
	}

	// Sort edges by cost in ascending order.
	sort.Slice(c.Edges, func(i, j int) bool {
		return c.Edges[i].Cost < c.Edges[j].Cost
	})

	uf := NewUnionFind(c.Nodes) // Initialize the Union-Find structure.
	numClusters := c.Nodes      // Start with each node as its own cluster.

	// Process edges to form clusters.
	for _, edge := range c.Edges {
		// Check if the nodes belong to different clusters.
		if uf.Find(edge.Node1) != uf.Find(edge.Node2) {
			if numClusters == k {
				// If the desired number of clusters is reached, return the next edge cost.
				return edge.Cost, nil
			}
			uf.Union(edge.Node1, edge.Node2) // Merge the clusters.
			numClusters--                    // Decrease the number of clusters.
		}
	}

	return 0, fmt.Errorf("unable to determine max spacing")
}

// ReadInput reads the input data from a file and returns the nodes and edges.
func ReadInput(filePath string) (int, []Edge, error) {
	file, err := os.Open(filePath) // Open the input file.
	if err != nil {
		return 0, nil, fmt.Errorf("unable to open file: %w", err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	if !scanner.Scan() {
		return 0, nil, fmt.Errorf("file is empty or missing number of nodes")
	}

	numNodes, err := strconv.Atoi(scanner.Text()) // Read the number of nodes.
	if err != nil {
		return 0, nil, fmt.Errorf("invalid number of nodes: %w", err)
	}

	var edges []Edge
	lineNumber := 1
	// Read each line as an edge.
	for scanner.Scan() {
		lineNumber++
		parts := strings.Fields(scanner.Text())
		if len(parts) != 3 {
			return 0, nil, fmt.Errorf("invalid line %d: expected 3 fields, got %d", lineNumber, len(parts))
		}

		node1, err := strconv.Atoi(parts[0])
		if err != nil {
			return 0, nil, fmt.Errorf("invalid node1 at line %d: %w", lineNumber, err)
		}

		node2, err := strconv.Atoi(parts[1])
		if err != nil {
			return 0, nil, fmt.Errorf("invalid node2 at line %d: %w", lineNumber, err)
		}

		cost, err := strconv.Atoi(parts[2])
		if err != nil {
			return 0, nil, fmt.Errorf("invalid cost at line %d: %w", lineNumber, err)
		}

		// Add the edge to the list.
		edges = append(edges, Edge{Node1: node1 - 1, Node2: node2 - 1, Cost: cost})
	}

	if err := scanner.Err(); err != nil {
		return 0, nil, fmt.Errorf("error reading file: %w", err)
	}

	return numNodes, edges, nil
}

func main() {
	// Define the input file path.
	filePath := "course_3/module_2/programming_assignment_2/clustering1.txt"

	// Read input data from the file.
	nodes, edges, err := ReadInput(filePath)
	if err != nil {
		fmt.Printf("Error reading input file: %v\n", err)
		return
	}

	// Initialize the clustering algorithm.
	clustering := NewClustering(nodes, edges)

	// Compute the maximum spacing for 4 clusters.
	maxSpacing, err := clustering.ComputeMaxSpacing(4)
	if err != nil {
		fmt.Printf("Error computing max spacing: %v\n", err)
		return
	}

	// Print the result.
	fmt.Printf("The maximum spacing of a 4-clustering is: %d\n", maxSpacing)
}
